Mathlib/Topology/Order/UpperSetTopology.lean

```diff
@@ -0,0 +1,283 @@
+  right_inv := toUpperSet_ofUpperSet,
+  map_rel_iff' := toUpperSet_le_iff }
+
+end WithUpperSetTopology
+
+/--
+The upper set topology is the topology where the open sets are the upper sets. In general the upper
+set topology does not coincide with the dual of the lower topology.
+-/
+class UpperSetTopology (α : Type _) [t : TopologicalSpace α] [Preorder α] : Prop where
+  topology_eq_upperSetTopology : t = upperSetTopology' α
+
+instance [Preorder α] : UpperSetTopology (WithUpperSetTopology α) :=
+  ⟨rfl⟩
+
+instance [Preorder α] : @UpperSetTopology (WithUpperSetTopology α) (upperSetTopology' α) _ := ⟨rfl⟩
+
+instance [Preorder α] : @UpperSetTopology α (upperSetTopology' α) _ := by
+  letI := upperSetTopology' α
+  exact ⟨rfl⟩
+
+
+namespace UpperSetTopology
+
+section Preorder
+
+variable (α)
+variable [Preorder α] [TopologicalSpace α] [UpperSetTopology α] {s : Set α}
+
+lemma topology_eq : ‹_› = upperSetTopology' α := topology_eq_upperSetTopology
+
+variable {α}
+
+/-- If `α` is equipped with the upper set topology, then it is homeomorphic to
+`WithUpperSetTopology α`.
+-/
+def withUpperSetTopologyHomeomorph : WithUpperSetTopology α ≃ₜ α :=
+  WithUpperSetTopology.ofUpperSet.toHomeomorphOfInducing ⟨by erw [topology_eq α, induced_id]; rfl⟩
+
+lemma IsOpen_iff_IsUpperSet : IsOpen s ↔ IsUpperSet s := by
+  rw [topology_eq α]
+  rfl
+
+-- Alexandrov property, set formulation
+theorem IsOpen_sInter {S : Set (Set α)} (hf : ∀ s ∈ S, IsOpen s) : IsOpen (⋂₀ S) := by
+  rw [IsOpen_iff_IsUpperSet]
+  apply isUpperSet_sInter
+  intros s hs
+  rw [← IsOpen_iff_IsUpperSet]
+  exact hf _ hs
+
+-- Alexandrov property, index formulation
+theorem isOpen_iInter {f : ι → Set α} (hf : ∀ i, IsOpen (f i)) : IsOpen (⋂ i, f i) := by
+  rw [IsOpen_iff_IsUpperSet]
+  apply isUpperSet_iInter
+  intros i
+  rw [← IsOpen_iff_IsUpperSet]
+  exact hf i
+
+-- c.f. isClosed_iff_lower_and_subset_implies_LUB_mem
+lemma isClosed_iff_isLower {s : Set α} : IsClosed s ↔ (IsLowerSet s) := by
+  rw [← isOpen_compl_iff, IsOpen_iff_IsUpperSet,
+    isLowerSet_compl.symm, compl_compl]
+
+lemma isClosed_isLower {s : Set α} : IsClosed s → IsLowerSet s := fun h =>
+  (isClosed_iff_isLower.mp h)
+
+lemma closure_eq_lowerClosure {s : Set α} : closure s = lowerClosure s := by
+  rw [subset_antisymm_iff]
+  constructor
+  · apply closure_minimal subset_lowerClosure _
+    rw [isClosed_iff_isLower]
+    exact LowerSet.lower (lowerClosure s)
+  · apply lowerClosure_min subset_closure (isClosed_isLower isClosed_closure)
+
+/--
+The closure of a singleton `{a}` in the upper set topology is the right-closed left-infinite
+interval (-∞,a].
+-/
+@[simp] lemma closure_singleton {a : α} : closure {a} = Iic a := by
+  rw [closure_eq_lowerClosure, lowerClosure_singleton]
+  rfl
+
+end Preorder
+
+section maps
+
+variable [Preorder α] [Preorder β]
+
+lemma upperSetTopology_coinduced {t₁ : TopologicalSpace α} [UpperSetTopology α]
+    (hf : Monotone f) : coinduced f t₁ ≤ upperSetTopology' β := by
```

Comment:
Is there a particular reason you split out this result from the TFAE? Ease of use for defining the Scott topology?

Reply:
I think I used it in that form here: https://github.com/leanprover-community/mathlib4/pull/2508/files#diff-983e16dd03108ad09234625f776ad94cfb58b1f6ee4762bf7519306fd90a2b46R220

To be honest, I'm finding multiple topologies on the same space in Lean very hard to wrap my head around. Given that it's a very common situation in mathematics (think about all the topologies on a W*-algebra - e.g. Takesaki, Theory of Operator Algebra I Chapter II Theorem 2.6) I was hoping there might be more by the way of an established convention about how to set these things out. I asked [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Multiple.20topologies.3F) but found it hard to interpret the answers.

Reply:
I didn't participate in this thread at the time because I was busy with other things, but it's true that the discussion wasn't super clear. I have quite a lot of experience working with multiple topologies (although not in the context of order-related topologies). I'll try to write a summary of the way I see things tomorrow and post it on Zulip. 

I also have ideas about this particular situation, but I need to test them so I'll comment tomorrow.

Reply:
I've added some explanations on Zulip, I hope that'll help to clarify things!